[jit] Consider contained types in alias analysis

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21431 [jit] Consider contained types in alias analysis**
* #21430 [jit] cleanups to memory_dag
* #21397 [jit] cleanups to alias analysis interfaces
* #21396 [jit] avoid calling front() on empty working set

`AliasDb::getReads()` now considers contained types to be part of the
read set.

This is necessary because we need to consider reads of List[Tensor] to
be reads both of the outer List and the inner Tensor. Otherwise
something like this might happen:
```
l = [...]
l[0].add_(...)
print(l)  # could reorder this read across the write above
```

This can be improved (e.g. we can track tuple contained aliasing
precisely) but I will defer that to the future.

gh-metadata: pytorch pytorch 21431 gh/suo/53/head

Differential Revision: [D15718560](https://our.internmc.facebook.com/intern/diff/D15718560)

suo committed:
[jit] Track contained elements in alias analysis.

Container types now track the aliasing information of their contained
elements. Creating a pointer from one container to another should
correctly create a pointer between their contained types as well.

This is necessary because we need to consider reads of List[Tensor] to
be reads both of the outer List and the inner Tensor. Otherwise
something like this might happen:
```
l = [...]
l[0].add_(...)
print(l)  # could reorder this read across the write above
```

I think this change also makes it so that we can be more explicit with
schema annotations (wildcards can become implementation details
finally!) A follow-on PR will try to make that happen

suo commented:
actually this is not ready for review yet

suo committed:
Update on "[jit] Track contained elements in alias analysis."

[jit] Track contained elements in alias analysis.

Container types now track the aliasing information of their contained
elements. Creating a pointer from one container to another should
correctly create a pointer between their contained types as well.

This is necessary because we need to consider reads of List[Tensor] to
be reads both of the outer List and the inner Tensor. Otherwise
something like this might happen:
```
l = [...]
l[0].add_(...)
print(l)  # could reorder this read across the write above
```

I think this change also makes it so that we can be more explicit with
schema annotations (wildcards can become implementation details
finally!) A follow-on PR will try to make that happen

gh-metadata: pytorch pytorch 21431 gh/suo/53/head

suo committed:
Update on "[jit] Track contained elements in alias analysis."

[jit] Track contained elements in alias analysis.

Container types now track the aliasing information of their contained
elements. Creating a pointer from one container to another should
correctly create a pointer between their contained types as well.

This is necessary because we need to consider reads of List[Tensor] to
be reads both of the outer List and the inner Tensor. Otherwise
something like this might happen:
```
l = [...]
l[0].add_(...)
print(l)  # could reorder this read across the write above
```

I think this change also makes it so that we can be more explicit with
schema annotations (wildcards can become implementation details
finally!) A follow-on PR will try to make that happen

gh-metadata: pytorch pytorch 21431 gh/suo/53/head

suo committed:
Update on "[jit] Track contained elements in alias analysis."

[jit] Track contained elements in alias analysis.

Container types now track the aliasing information of their contained
elements. Creating a pointer from one container to another should
correctly create a pointer between their contained types as well.

This is necessary because we need to consider reads of List[Tensor] to
be reads both of the outer List and the inner Tensor. Otherwise
something like this might happen:
```
l = [...]
l[0].add_(...)
print(l)  # could reorder this read across the write above
```

I think this change also makes it so that we can be more explicit with
schema annotations (wildcards can become implementation details
finally!) A follow-on PR will try to make that happen

gh-metadata: pytorch pytorch 21431 gh/suo/53/head

suo committed:
Update on "[jit] Track contained elements in alias analysis."

[jit] Track contained elements in alias analysis.

Container types now track the aliasing information of their contained
elements. Creating a pointer from one container to another should
correctly create a pointer between their contained types as well.

This is necessary because we need to consider reads of List[Tensor] to
be reads both of the outer List and the inner Tensor. Otherwise
something like this might happen:
```
l = [...]
l[0].add_(...)
print(l)  # could reorder this read across the write above
```

I think this change also makes it so that we can be more explicit with
schema annotations (wildcards can become implementation details
finally!) A follow-on PR will try to make that happen

gh-metadata: pytorch pytorch 21431 gh/suo/53/head

zdevito commented:
I am going to need a refresher on our alias analysis pass before I have enough context to review for correctness.

suo committed:
Update on "[jit] Consider contained types in alias analysis"

[jit] Consider contained types in alias analysis

`AliasDb::getReads()` now considers contained types to be part of the
read set.

This is necessary because we need to consider reads of List[Tensor] to
be reads both of the outer List and the inner Tensor. Otherwise
something like this might happen:
```
l = [...]
l[0].add_(...)
print(l)  # could reorder this read across the write above
```

This can be improved (e.g. we can track tuple contained aliasing
precisely) but I will defer that to the future.

gh-metadata: pytorch pytorch 21431 gh/suo/53/head

suo commented:
@zdevito: Changed this as we discussed offline. Not doing any of the container elements stuff—just adding the appropriate wildcards to the `getReads()` set.

suo committed:
Update on "[jit] Consider contained types in alias analysis"

[jit] Consider contained types in alias analysis

`AliasDb::getReads()` now considers contained types to be part of the
read set.

This is necessary because we need to consider reads of List[Tensor] to
be reads both of the outer List and the inner Tensor. Otherwise
something like this might happen:
```
l = [...]
l[0].add_(...)
print(l)  # could reorder this read across the write above
```

This can be improved (e.g. we can track tuple contained aliasing
precisely) but I will defer that to the future.

gh-metadata: pytorch pytorch 21431 gh/suo/53/head

suo committed:
Update on "[jit] Consider contained types in alias analysis"

[jit] Consider contained types in alias analysis

`AliasDb::getReads()` now considers contained types to be part of the
read set.

This is necessary because we need to consider reads of List[Tensor] to
be reads both of the outer List and the inner Tensor. Otherwise
something like this might happen:
```
l = [...]
l[0].add_(...)
print(l)  # could reorder this read across the write above
```

This can be improved (e.g. we can track tuple contained aliasing
precisely) but I will defer that to the future.

gh-metadata: pytorch pytorch 21431 gh/suo/53/head

suo committed:
Update on "[jit] Consider contained types in alias analysis"

[jit] Consider contained types in alias analysis

`AliasDb::getReads()` now considers contained types to be part of the
read set.

This is necessary because we need to consider reads of List[Tensor] to
be reads both of the outer List and the inner Tensor. Otherwise
something like this might happen:
```
l = [...]
l[0].add_(...)
print(l)  # could reorder this read across the write above
```

This can be improved (e.g. we can track tuple contained aliasing
precisely) but I will defer that to the future.

gh-metadata: pytorch pytorch 21431 gh/suo/53/head
